File: src/main/java/org/cerion/stocklist/overlays/PriceOverlayBase.java
package org.cerion.stocklist.overlays;

import org.cerion.stocklist.PriceList;
import org.cerion.stocklist.arrays.ValueArray;
import org.cerion.stocklist.functions.FunctionBase;
import org.cerion.stocklist.functions.types.IFunctionEnum;
import org.cerion.stocklist.functions.IPriceOverlay;

import java.lang.reflect.Method;

abstract class PriceOverlayBase extends FunctionBase implements IPriceOverlay {

    PriceOverlayBase(IFunctionEnum id, Number ...params) {
        super(id, params);
    }

    @Override
    @SuppressWarnings("unchecked")
    public Class<? extends ValueArray> getResultType() {
        try {
            Method evalMethod = getClass().getMethod("eval", PriceList.class);
            return (Class<? extends ValueArray>)evalMethod.getReturnType();
        }
        catch (Exception e){
            throw new RuntimeException(e.getMessage());
        }
    }
}

File: src/main/java/org/cerion/stocklist/overlays/PriceOverlayBase.kt
package org.cerion.stocklist.overlays

import org.cerion.stocklist.PriceList
import org.cerion.stocklist.arrays.ValueArray
import org.cerion.stocklist.functions.FunctionBase
import org.cerion.stocklist.functions.IPriceOverlay
import org.cerion.stocklist.functions.types.IFunctionEnum

abstract class PriceOverlayBase internal constructor(id: IFunctionEnum, vararg params: Number) : FunctionBase(id, *params), IPriceOverlay {

    override fun getResultType(): Class<out ValueArray> {
        try {
            val evalMethod = javaClass.getMethod("eval", PriceList::class.java)
            return evalMethod.returnType as Class<out ValueArray>
        } catch (e: Exception) {
            throw RuntimeException(e.message)
        }
    }
}
